module/bdev/raid/bdev_raid.c

```diff
@@ -1749,15 +1801,302 @@ raid_bdev_remove_base_bdev(struct spdk_bdev *base_bdev)
 			raid_bdev_cleanup_and_free(raid_bdev);
 		}
+		raid_bdev->base_bdev_updating = false;
 	} else if (raid_bdev->num_base_bdevs_operational-- == raid_bdev->min_base_bdevs_operational) {
 		raid_bdev_deconfigure(raid_bdev, NULL, NULL);
+		raid_bdev->base_bdev_updating = false;
 	} else {
 		return raid_bdev_suspend(raid_bdev, raid_bdev_remove_base_bdev_on_suspended, base_info);
 	}
 
 	return 0;
 }
 
+struct raid_bdev_add_base_bdev_ctx {
+	struct raid_bdev *raid_bdev;
+	char *base_bdev_name;
+	/* Status is the result of the operation, 0 if successful otherwise -ERRNO */
+	int status;
+	uint8_t slot;
+	raid_bdev_destruct_cb cb_fn;
+	void *cb_arg;
+};
+
+static void
+raid_bdev_add_base_bdev_on_resumed(void *_ctx, int rc)
+{
+	struct raid_bdev_add_base_bdev_ctx *ctx = _ctx;
+
+	if (rc && !ctx->status) {
+		ctx->status = rc;
+	}
+
+	ctx->raid_bdev->base_bdev_updating = false;
+	ctx->cb_fn(ctx->cb_arg, ctx->status);
+	free(ctx->base_bdev_name);
+	free(ctx);
+}
+
+static void
+raid_bdev_add_base_bdev_write_sb_cb(bool success, struct raid_bdev *raid_bdev, void *_ctx)
+{
+	struct raid_bdev_add_base_bdev_ctx *ctx = _ctx;
+
+	if (!success) {
+		SPDK_ERRLOG("Failed to write raid bdev '%s' superblock\n", raid_bdev->bdev.name);
+		ctx->status = -EIO;
+	}
+
+	raid_bdev_resume(raid_bdev, raid_bdev_add_base_bdev_on_resumed, ctx);
+}
+
+static void
+raid_bdev_add_base_bdev_done(struct spdk_io_channel_iter *i, int status)
+{
+	struct raid_bdev_add_base_bdev_ctx *ctx = spdk_io_channel_iter_get_ctx(i);
+	struct raid_bdev *raid_bdev = ctx->raid_bdev;
+	struct raid_base_bdev_info *base_info = &raid_bdev->base_bdev_info[ctx->slot];
+
+	if (status && !ctx->status) {
+		ctx->status = status;
+	}
+
+	if (!ctx->status && raid_bdev->sb) {
+		struct raid_bdev_superblock *sb = raid_bdev->sb;
+		struct raid_bdev_sb_base_bdev *sb_base_bdev;
+		int rc;
+		uint8_t i;
+
+		/* Check if we are filling ah hole */
+		for (i = 0; i < sb->base_bdevs_size; i++) {
+			sb_base_bdev = &sb->base_bdevs[i];
+
+			if (sb_base_bdev->state == RAID_SB_BASE_BDEV_REMOVED &&
+			    sb_base_bdev->slot == ctx->slot) {
+				break;
+			}
+		}
+
+		/* New slot */
+		if (i == sb->base_bdevs_size) {
+			sb_base_bdev = &sb->base_bdevs[i];
+
+			/* Update superblock num base bdevs and length */
+			sb->num_base_bdevs = sb->base_bdevs_size = raid_bdev->num_base_bdevs;
+			sb->length = sizeof(*sb) + sizeof(*sb_base_bdev) * sb->base_bdevs_size;
+		}
+
+		spdk_uuid_copy(&sb_base_bdev->uuid, spdk_bdev_get_uuid(base_info->bdev));
+		sb_base_bdev->data_offset = base_info->data_offset;
+		sb_base_bdev->data_size = base_info->data_size;
+		sb_base_bdev->state = RAID_SB_BASE_BDEV_CONFIGURED;
+		sb_base_bdev->slot = sb_base_bdev - sb->base_bdevs;
+
+		rc = raid_bdev_write_superblock(raid_bdev, raid_bdev_add_base_bdev_write_sb_cb, ctx);
+		if (rc != 0) {
+			raid_bdev_add_base_bdev_write_sb_cb(false, raid_bdev, ctx);
+		}
+	} else {
+		raid_bdev_resume(raid_bdev, raid_bdev_add_base_bdev_on_resumed, ctx);
```

Comment:
Why not

```suggestion
		rc = raid_bdev_write_superblock(raid_bdev, raid_bdev_add_base_bdev_write_sb_cb, ctx);
		if (rc == 0) {
			return;
		}
	}
	raid_bdev_resume(raid_bdev, raid_bdev_add_base_bdev_on_resumed, ctx);
```

Reply:
Thanks to this suggestion I found that, in case of `ENOMEM` error, `raid_bdev_write_superblock` doesn't call the callback function. In all other cases `raid_bdev_add_base_bdev_write_sb_cb` is called and consequentially the `raid_bdev_resume`.

Reply:
Got it. Thanks.